### PR TITLE
Test for Unet

### PR DESCRIPTION
**Pull request by mmaksimovicTT**
### Ticket
Linked to [tt-xla issue #322](https://github.com/tenstorrent/tt-xla/issues/322)

### Problem description
Adds a test for the UNet model, which currently fails due to missing decomposition from ttir.convolution to ttir.conv_transpose2d.
Tracked in [tt-xla issue #417](https://github.com/tenstorrent/tt-xla/issues/417)

### CODE REVIEW

**Comment by ajakovljevicTT** [comment]
Please move the tests to `tests/jax/single_chip/models`

File: tests/jax/single_chip/models/UNet/model_implementation.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+            )
+            up_blocks.append((upconv, double_conv))
+            in_ch = out_ch
+        self.up_sampling_blocks = tuple(up_blocks)
+
+        self.output_conv = nn.Conv(
+            features=self.out_channels, kernel_size=(1, 1), strides=(1, 1)
+        )
+
+    @nn.compact
```
Comment (sgligorijevicTT) [inline tests/jax/single_chip/models/UNet/model_implementation.py:71]: I didn't know it possible to use both `nn.compact` and a separate `setup` function together. You're supposed to use one or the other. I would remove `compact`, it looks like you aren't relying on it anyway.

File: tests/jax/single_chip/models/UNet/model_implementation.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        for block in self.down[:-1]:
+            x = nn.max_pool(x, window_shape=(2, 2), strides=(2, 2), padding="VALID")
+            x = block(x, train)
+            skips.append(x)
+        x = nn.max_pool(x, window_shape=(2, 2), strides=(2, 2), padding="VALID")
+        x = self.down[-1](x, train)
+
+        # After upsampling with ConvTranspose, 'x' might not perfectly match the size of the corresponding skip connection 'skip'.
+        # To concatenate them correctly, we center-crop 'skip' to match 'x' spatially.
+        # Then, we concatenate 'x' and cropped 'skip' along the channel axis before applying DoubleConv.
```
Comment (kmitrovicTT) [inline tests/jax/single_chip/models/UNet/model_implementation.py:91]: Great explanation.

Comment formatted terribly. Try ty line sentences up to 88chars. Set vscode ruler to 88chars. Comments should look something like

image

File: tests/jax/single_chip/models/UNet/model_implementation.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    def __call__(self, x: jnp.ndarray, train: bool) -> jnp.ndarray:
+        skips = []
+        x = self.lifting(x, train)
+        skips.append(x)
+
+        for block in self.down[:-1]:
+            x = nn.max_pool(x, window_shape=(2, 2), strides=(2, 2), padding="VALID")
+            x = block(x, train)
+            skips.append(x)
+        x = nn.max_pool(x, window_shape=(2, 2), strides=(2, 2), padding="VALID")
```
Comment (kmitrovicTT) [inline tests/jax/single_chip/models/UNet/model_implementation.py:86]: Please put blank lines after `for` blocks finish.

**Review by kmitrovicTT — APPROVED** [review]
I approve it, but I didn't check the implementation. @mrakitaTT will you be taking a look?

File: tests/jax/single_chip/models/UNet/model_implementation.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+            x = nn.max_pool(x, window_shape=(2, 2), strides=(2, 2), padding="VALID")
+            x = block(x, train)
+            skips.append(x)
+
+        x = nn.max_pool(x, window_shape=(2, 2), strides=(2, 2), padding="VALID")
+        x = self.down[-1](x, train)
+
+        """
+        After upsampling with ConvTranspose, 'x' might not perfectly match the size
+        of the corresponding skip connection. To concatenate them correctly, we
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/UNet/model_implementation.py:94]: I would add this:
> After upsampling with ConvTranspose, 'x' **will** not match the size of the corresponding skip connection **because the convolutions are applied unpadded which results in lower image size after each convolution.**

File: tests/jax/single_chip/models/UNet/model_implementation.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        for i in range(self.num_levels):
+            x = self.up_convs[i](x)
+            skip = skips[-(i + 1)]
+            diff_h = skip.shape[1] - x.shape[1]
+            diff_w = skip.shape[2] - x.shape[2]
+            start_h = diff_h // 2
+            start_w = diff_w // 2
+            skip = skip[
+                :, start_h : start_h + x.shape[1], start_w : start_w + x.shape[2], :
+            ]
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/UNet/model_implementation.py:107]: I would extract this code into separate function `crop_skip_connection(skip_connection, crop_shape)`, and do here `skip_connection = crop_skip_connection(skip_connection, x.shape)`

File: tests/jax/single_chip/models/UNet/model_implementation.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+            in_ch = out_ch
+
+        self.up_blocks = tuple(up_blocks)
+        self.up_convs = tuple(up_convs)
+
+        self.output_conv = nn.Conv(
+            self.out_channels, kernel_size=(1, 1), strides=(1, 1), padding="VALID"
+        )
+
+    def __call__(self, x: jnp.ndarray, train: bool) -> jnp.ndarray:
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/UNet/model_implementation.py:79]: After you make all the changes, in order to verify that everything is still fine you can add prints after each step to print the shape of the `x`, and then verify that it matches shapes from the illustration in the paper. These prints are just for your verification, shouldn't add them to PR.

File: tests/jax/single_chip/models/UNet/model_implementation.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+
+class DoubleConv(nn.Module):
+    in_channels: int
+    out_channels: int
+    activation: Callable = nn.relu
+
+    def setup(self):
+        self.conv1 = nn.Conv(
+            self.out_channels, kernel_size=(3, 3), padding="VALID", use_bias=False
```
[the remark this answers is not in the document]
Comment (mmaksimovicTT) [reply tests/jax/single_chip/models/UNet/model_implementation.py:17]: In the original U-Net paper, they did use use_bias=True since there was no BatchNorm. But in my version, I added BatchNorm — and when you use it, the conv bias becomes unnecessary because BatchNorm already learns a bias term. So I set use_bias=False to avoid redundancy.

File: tests/jax/single_chip/models/UNet/model_implementation.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        )
+        self.bn1 = nn.BatchNorm()
```
[the remark this answers is not in the document]
Comment (mmaksimovicTT) [reply tests/jax/single_chip/models/UNet/model_implementation.py:19]: Thanks for the comment — you're right, I should've added a flag and defaulted it to False for consistency with the original paper.
I kept BatchNorm to make PyTorch-to-JAX weight conversion easier — functionality stays the same.

File: tests/jax/single_chip/models/unet/model_implementation.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+
+class DoubleConv(nn.Module):
+    in_channels: int
+    out_channels: int
+    activation: Callable = nn.relu
+    use_batchnorm: bool = False
+    param_dtype: Optional[
+        Union[str, type[Any], jnp.dtype, jax._src.typing.SupportsDType, Any]
+    ] = jnp.bfloat16
```
Comment (kmitrovicTT) [inline tests/jax/single_chip/models/unet/model_implementation.py:19]: `from jax._src.typing import DTypeLike` should cover you on this.

**Review by kmitrovicTT — APPROVED** [review]

File: tests/jax/single_chip/models/unet/model_implementation.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+
+    def setup(self):
+        self.upconv = nn.ConvTranspose(
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/unet/model_implementation.py:88]: Nit: rename `self.upconv` to `self.up_conv`, to be consistent with `double_conv`

File: tests/jax/single_chip/models/unet/model_implementation.py [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+            self.hidden_channels,
+            activation=self.activation,
+            use_batchnorm=self.use_batchnorm,
+            param_dtype=self.param_dtype,
+        )
+
+        down_blocks = []
+        in_ch = self.hidden_channels
+        for i in range(self.num_levels):
+            out_ch = self.hidden_channels * 2 ** (i + 1)
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/unet/model_implementation.py:154]: `out_ch = in_ch * 2`

File: tests/jax/single_chip/models/unet/model_implementation.py [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+                    in_channels=in_ch,
+                    out_channels=out_ch,
+                    activation=self.activation,
+                    use_batchnorm=self.use_batchnorm,
+                    param_dtype=self.param_dtype,
+                )
+            )
+            in_ch = out_ch
+
+        in_ch = self.hidden_channels * 2**self.num_levels
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/unet/model_implementation.py:166]: I think this line is not necessary, the last `in_ch = out_ch` will already set `in_ch` to this value. Please check.

File: tests/jax/single_chip/models/unet/model_implementation.py [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+    def setup(self):
+        self.upconv = nn.ConvTranspose(
+            features=self.out_channels,
+            kernel_size=(2, 2),
+            strides=(2, 2),
+            padding="VALID",
+            param_dtype=self.param_dtype,
+        )
+        self.double_conv = DoubleConv(
+            in_channels=self.out_channels * 2,
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/unet/model_implementation.py:96]: Should use `self.in_channels` here

File: tests/jax/single_chip/models/unet/model_implementation.py [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+                    param_dtype=self.param_dtype,
+                )
+            )
+            in_ch = out_ch
+
+        in_ch = self.hidden_channels * 2**self.num_levels
+
+        up_blocks = []
+        for i in range(self.num_levels):
+            out_ch = self.hidden_channels * 2 ** (self.num_levels - i - 1)
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/unet/model_implementation.py:170]: `out_ch = in_ch // 2`

File: tests/jax/single_chip/models/unet/model_implementation.py (hunk 13 above)
Comment (mrakitaTT) [reply tests/jax/single_chip/models/unet/model_implementation.py:96]: I've just realised that you are not actually using `in_channels` parameter anywhere, neither in `DoubleConv` nor in `UpBlock` and `DownBlock` :) You can remove it from all components and just have the `out_channels` parameter.

File: tests/jax/single_chip/models/unet/model_implementation.py [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+        skip_connections = []
+
+        x = self.lifting(x, train)
+        skip_connections.append(x)
+
+        for block in self.down_blocks:
+            x = block(x, train)
+            skip_connections.append(x)
+
+        skip_connections = skip_connections[:-1]
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/unet/model_implementation.py:203]: Your `DownBlock` implementation simplifies setting skip connections, but one negative consequence is that we are now calculating one skip connection extra, i.e. `crop_skip_connection` is being called one extra time, wasting computation time and memory. That is why I've suggested [here](https://github.com/tenstorrent/tt-xla/pull/422#discussion_r2067417617) to reverse the order and make `DownBlock` in such a way that it first applies `DoubleConv` and then `nn.max_pool`, avoiding need for separate `self.lifting` block and avoiding creating one extra skip connection. What do you think about that approach?

File: tests/jax/single_chip/models/unet/model_implementation.py [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+
+from typing import Any, Callable, Optional, Union
+
+import flax.linen as nn
+import jax.numpy as jnp
+from jax._src.typing import DTypeLike
+
+
+class DoubleConv(nn.Module):
+    in_channels: int
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/unet/model_implementation.py:13]: `in_channels` is not being used in the `DoubleConv` class, can be removed

File: tests/jax/single_chip/models/unet/model_implementation.py [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+        x = self.conv2(x)
+        if self.use_batchnorm:
+            x = self.bn2(x, use_running_average=not train)
+        x = self.activation(x)
+
+        return x
+
+
+class DownBlock(nn.Module):
+    in_channels: int
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/unet/model_implementation.py:53]: Here too, it is used only in `DoubleConv` constructor but it will be removed from `DoubleConv` too

File: tests/jax/single_chip/models/unet/model_implementation.py [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+    def __call__(self, x, train: bool):
+        feature_map = self.double_conv(x, train)
+        pooled_x = nn.max_pool(
+            feature_map, window_shape=(2, 2), strides=(2, 2), padding="VALID"
+        )
+        return pooled_x, feature_map
+
+
+class UpBlock(nn.Module):
+    in_channels: int
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/unet/model_implementation.py:77]: Here too

File: tests/jax/single_chip/models/unet/model_implementation.py [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+            down_blocks.append(
+                DownBlock(
+                    in_channels=in_ch,
+                    out_channels=out_ch,
+                    activation=self.activation,
+                    use_batchnorm=self.use_batchnorm,
+                    param_dtype=self.param_dtype,
+                )
+            )
+            in_ch = out_ch
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/unet/model_implementation.py:150]: You are not using `hidden_channels` here. Should be like this:
```
        out_ch = self.in_channels * self. hidden_channels
        for i in range(self.num_levels):
            down_blocks.append(
                DownBlock(
                    out_channels=out_ch,
                    activation=self.activation,
                    use_batchnorm=self.use_batchnorm,
                    param_dtype=self.param_dtype,
                )
            )
            out_ch = out_ch * 2
```

File: tests/jax/single_chip/models/unet/model_implementation.py [hunk 20]
```diff
@@ -0,0 +1,199 @@
...
+
+        self.bottom_double_conv = DoubleConv(
+            in_channels=in_ch,
+            out_channels=in_ch * 2,
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/unet/model_implementation.py:154]: After that `out_ch` will already be set to 1024 so here you can do just `out_channels=out_ch`

File: tests/jax/single_chip/models/unet/model_implementation.py [hunk 21]
```diff
@@ -0,0 +1,199 @@
...
+            activation=self.activation,
+            use_batchnorm=self.use_batchnorm,
+            param_dtype=self.param_dtype,
+        )
+
+        up_blocks = []
+        in_ch = in_ch * 2
+        for i in range(self.num_levels):
+            out_ch = in_ch // 2
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/unet/model_implementation.py:163]: And then here you just do `out_ch = out_ch // 2` and remove the `in_ch` lines

File: tests/jax/single_chip/models/unet/model_implementation.py [hunk 22]
```diff
@@ -0,0 +1,199 @@
...
+
+        self.output_conv = nn.Conv(
+            self.out_channels,
+            kernel_size=(1, 1),
+            strides=(1, 1),
+            padding="VALID",
+            param_dtype=self.param_dtype,
+        )
+
+    def __call__(self, x: jnp.ndarray, train: bool) -> jnp.ndarray:
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/unet/model_implementation.py:186]: Just a debug advice, I would copy this model implementation into separate script and run it in eager mode (without jit), so you can print the shape of x after each step here, and in `__call__` function of other blocks if needed, to verify that you are getting the expected shapes.
